Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -1864,30 +1861,30 @@ private static RelativePath GetHashSubDirectory(ContentHash contentHash)
 
         internal bool TryGetFileInfo(ContentHash contentHash, out ContentFileInfo fileInfo) => ContentDirectory.TryGetFileInfo(contentHash, out fileInfo);
 
-        internal List<(ContentHash hash, FileInfo fileInfo)> ReadContentHashesFromDisk(Context context)
+        internal ContentHashAddressableSnapshot<FileInfo> ReadSnapshotFromDisk(Context context)
         {
-            var contentHashes = new List<(ContentHash hash, FileInfo fileInfo)>();
+            var contentHashes = new List<PayloadFromDisk<FileInfo>>();
             if (_settings.UseNativeBlobEnumeration)
             {
-                EnumerateBlobPathsFromDisk(context, fileInfo => parseAndAccumulateContentHashes(fileInfo));
+                EnumerateBlobPathsFromDisk(context, fileInfo => ParseAndAccumulateContentHashes(fileInfo));
             }
             else
             {
                 foreach (var fileInfo in EnumerateBlobPathsFromDisk())

```

Comment:
>EnumerateBlobPaths [](start = 41, length = 18)

Please add comments here (and elsewhere where we construct list from enumeration) describing that we were hit by large object limits, so we want to make the objects in the list reference types or small structures.